### PR TITLE
Adding a new data generation function based on logistic regression 

### PR DESCRIPTION
**Pull request by zhenyuz0500**
Adding a new data generation function for classification uplift problem.
This data generation function is based on logistic regression.
The advantages of this new function is to enable better control over feature importance and feature pattern. 
In the previous data generation function, if we specify M number of informative features, then in the end the actual number of informative features is m < M, many informative features behave similarly to irrelevant features. In this function, all informative features would show a clear pattern for impacting the outcome.
In addition, now we can specify the uplift feature patterns: the uplift feature pattern can be one of the following ['linear','quadratic','cubic','relu','sin','cos']. For example, if the feature pattern is 'quadratic', then the treatment effect will increase or decrease quadratically with the feature.

### CODE REVIEW

File: causalml/dataset/classification.py [hunk 1]
```diff
@@ -1,8 +1,426 @@
...
+    # generate conversion probability and true treatment effect ---------------------------------#
+    df1['conversion_prob'] = 1 / (1 + np.exp(- df1['conversion_prob_linear'].values))
+    df1['control_conversion_prob'] = 1 / (1 + np.exp(- df1['control_conversion_prob_linear'].values))
+    for treatment_key_i in treatment_name:
+        df1['%s_conversion_prob'%(treatment_key_i)] = 1 / (1 + np.exp(- df1['%s_conversion_prob_linear'%(treatment_key_i)].values))
+        df1['%s_true_effect'%(treatment_key_i)] = df1['%s_conversion_prob'%(treatment_key_i)].values - df1['control_conversion_prob'].values
+    
+    # generate Y ------------------------------------------------------------#
+    df1['conversion_prob'] = [max(0, min(1, xi)) 
+                              for xi in df1['conversion_prob'].values]
```
Comment (jeongyoonlee) [inline causalml/dataset/classification.py:415]: `df1['conversion_prob'] = np.clip(df1['conversion_prob'].values, 0, 1)`

File: causalml/dataset/classification.py [hunk 2]
```diff
@@ -1,8 +1,426 @@
...
+    Y1 = np.random.binomial(1, df1['conversion_prob'].values)
+
+    df1[y_name] = Y1
```
Comment (jeongyoonlee) [inline causalml/dataset/classification.py:418]: `Y1` is not necessary.

File: causalml/dataset/classification.py [hunk 3]
```diff
@@ -1,8 +1,426 @@
...
+ 
+
+
+
+    # generate conversion probability and true treatment effect ---------------------------------#
+    df1['conversion_prob'] = 1 / (1 + np.exp(- df1['conversion_prob_linear'].values))
+    df1['control_conversion_prob'] = 1 / (1 + np.exp(- df1['control_conversion_prob_linear'].values))
+    for treatment_key_i in treatment_name:
+        df1['%s_conversion_prob'%(treatment_key_i)] = 1 / (1 + np.exp(- df1['%s_conversion_prob_linear'%(treatment_key_i)].values))
+        df1['%s_true_effect'%(treatment_key_i)] = df1['%s_conversion_prob'%(treatment_key_i)].values - df1['control_conversion_prob'].values
```
Comment (jeongyoonlee) [inline causalml/dataset/classification.py:411]: We can use `scipy.special.expit()` here. e.g.:
```python
from scipy.special import expit

df1['conversion_prob'] = expit(df1['conversion_prob_linear'].values)
...
```

File: causalml/dataset/classification.py [hunk 4]
```diff
@@ -1,8 +1,426 @@
...
+            a20 = np.log(p2/(1.- p2)) - a1
+            xb_array = df1['conversion_prob_linear'].values + (x_uplift * coef_uplift).sum(axis=1) 
+            xb_array_treatment = xb_array[treatment_index]
+            a2 = fsolve(_softmax,a20,args=(p2,xb_array_treatment))[0]
+            df1['%s_conversion_prob_linear'%(treatment_key_i)] = a2 + xb_array
+            df1.loc[treatment_index,'conversion_prob_linear'] = df1.loc[treatment_index,'%s_conversion_prob_linear'%(treatment_key_i)].values
+        else:
+            df1['%s_conversion_prob_linear'%(treatment_key_i)] = df1['conversion_prob_linear'].values
+ 
+
```
Comment (jeongyoonlee) [inline causalml/dataset/classification.py:403]: Remove additional blank lines.

File: causalml/dataset/classification.py [hunk 5]
```diff
@@ -1,8 +1,426 @@
...
+        if treatment_key_i in delta_uplift_dict and np.abs(delta_uplift_dict[treatment_key_i]) > 0.:                 
+            treatment_index = (
+                df1.index[df1['treatment_group_key'] == treatment_key_i].tolist()
+            )
+            # coefficient 
+            coef_uplift = []
+            for ci in range(n_uplift_dict[treatment_key_i]):
+                #rcoef = [0]
+                #while np.abs(rcoef) < 0.1:
+                #    rcoef = np.random.uniform(-1, 1, 1)
```
Comment (jeongyoonlee) [inline causalml/dataset/classification.py:389]: Remove the commented code block

File: causalml/dataset/classification.py [hunk 6]
```diff
@@ -1,8 +1,426 @@
...
+                rcoef = [0.5]
+                coef_uplift.append(rcoef[0])
```
Comment (jeongyoonlee) [inline causalml/dataset/classification.py:391]: `rcoef` doesn't need to be an array.

File: causalml/dataset/classification.py [hunk 7]
```diff
@@ -1,8 +1,426 @@
...
+    Parameters
+    ----------
+    z : float
+        Constant value in the linear part.
+    p : float 
+        The target output mean value.
+    xb : list 
+        An array, with each element as the sum of product of coefficient and feature value 
+    """
+    sm_arr = 1/(1+np.exp(-(z+np.array(xb))))
```
Comment (jeongyoonlee) [inline causalml/dataset/classification.py:158]:
```python
from scipy.special import expit

sm_arr = expit(z + np.array(xb))
```

File: causalml/dataset/classification.py [hunk 8]
```diff
@@ -1,8 +1,426 @@
...
+    # baseline conversion 
+    coef_classify = []
+    for ci in range(n_classification_informative):
+        rcoef = [0]
+        while np.abs(rcoef) < 0.1:
+            rcoef = np.random.randn(1) * np.sqrt(1./n_classification_informative)
+        coef_classify.append(rcoef[0])
+    x_classify = df1[x_informative_transformed].values
+    p1 = positive_class_proportion
+    a10 = np.log(p1/(1.-p1))
```
Comment (jeongyoonlee) [inline causalml/dataset/classification.py:370]:
```python
from scipy.special import logit

a10 = logit(pi)
```

File: causalml/dataset/classification.py [hunk 9]
```diff
@@ -181,3 +599,8 @@ def make_uplift_classification(n_samples=1000,
     df_res[y_name] = Y
     df_res['treatment_effect'] = Y - Y1
     return df_res, x_name
+
+
+
+
+
```
Comment (jeongyoonlee) [inline causalml/dataset/classification.py:606]: Remove extra blank lines

**Comment by Patiljd** [comment]
For dgp

**Comment by rolandrmgservices** [comment]
@zhenyuz0500 @jeongyoonlee if this PR is still active, I'd be happy to work to address the review request and get it merged in.  I was able to check out the branch, build from source, and reproduce the notebook results with no issues.

**Comment by jeongyoonlee** [comment]
@rolandrmgservices, that'd be great. Thanks!

**Comment by jeongyoonlee** [comment]
Closing this PR in favor of #701.

**Pull request closed without merging**